New-app/new-build support for pipeline buildconfigs

jim-minter opened the pull request:
https://trello.com/c/H6ZwrpCo/995-5-new-app-new-build-support-for-pipeline-buildconfigs-techdebt

[test]

jim-minter commented:
@bparees Last test failed due to https://github.com/kubernetes/kubernetes/issues/32455 , but otherwise ptal

bparees commented on pkg/generate/app/cmd/resolve.go:
```diff
@@ -511,7 +525,18 @@ func AddMissingComponentsToRefBuilder(
+
+		case info.Jenkinsfile && (g.Strategy == generate.StrategyUnspecified || g.Strategy == generate.StrategyPipeline):
+			refs := b.AddComponents([]string{"scratch"}, func(input *app.ComponentInput) app.ComponentReference {
+				input.Resolver = app.FirstMatchResolver{Searcher: app.DockerClientSearcher{}}
```
why do we need a resolver here?

jim-minter replied:
There may well be a better way to do this.  I found the concept of a "scratch" component which doesn't seem to be much used, but was useful here because the code expects every repo to be attached to a component of some sort.  In the other strategy cases the component to use is clear; here I'm using "scratch" as a fake component.

I set the resolver to be fairly minimal but able to respond to the request for "scratch".  I wasn't clear that there was a neater way of doing this without a resolver but I'm all ears.

jim-minter commented:
@bparees thanks - all changes made as specified.

bparees commented on pkg/generate/app/cmd/resolve.go:
```diff
@@ -110,14 +115,18 @@ func Resolve(r *Resolvers, c *ComponentInputs, g *GenerationInputs) (*ResolvedCo
 
+	if g.BinaryBuild && g.Strategy == generate.StrategyPipeline {
+		return nil, errors.New("specifying binary builds and the pipeline strategy at the same time is not allowed")
+	}
```
can this check not be done in the cmd validation logic?  (newapp.CompleteAppConfig() ).  That seems like the more appropriate location.

jim-minter replied:
Done.

bparees commented:
lgtm for post 3.4

bparees commented:
failure looks legitimate:
```
--- FAIL: TestNewAppRunBuilds (4.90s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0xa0 pc=0x109bb1b]
```

bparees commented:
[merge]

wanghaoran1988 commented:
@jim-minter Do you think whether we need add a test scenario that the context dir of the repo contains both Dockerfile and Jenkinsfile, and I am not sure which kind of the build is our expected one.

bparees commented:
i think we decided the dockerfile should take precedence (retains backwards compatibility, sort of) and you use --strategy if you want a pipeline instead.

bparees commented:
(but yes the scenario should be tested)

jim-minter commented:
Hmm, in the case of unspecified strategy, I think what I've implemented is: if there's a Dockerfile, use it.  If not and there's a Jenkinsfile, use it.  Else if there's recognised bare source, use it (i.e. "Dockerfile > Jenkinsfile > bare source").  See `AddMissingComponentsToRefBuilder()`.

Previously it was: if there's a Dockerfile, use it.  Else if there's recognised bare source, use it (i.e. "Dockerfile > bare source").

I don't think that interposing Jenkinsfile in "Dockerfile > Jenkinsfile > bare source" can really be said to be maintaining backwards compatibility (it does for the Dockerfile case, but not for the bare source case).

I think that arguably we should go (have gone?) for "Jenkinsfile > Dockerfile > bare source" ("pipelines are now king") or perhaps "Dockerfile > bare source > Jenkinsfile" ("backwards compatibility is king").

I think that either is arguably preferential to what I implemented, but it's hard to choose between them and I'm not sure how much it matters to change it anyway. @bparees could you make a judgement call?

bparees commented:
I agree it's not clear what's best and no way to maintain true compatibility.

jenkinsfile -> dockerfile -> source seems reasonable to me. As you say, making jenkinsfile king makes sense.
